DelayedMessages/explainer.md

```diff
@@ -0,0 +1,757 @@
+  console.log(`message #${e.data.no}: started processing in worker at ${processingStartOnWorker.toFixed(2)} ms`);
+  console.log(
+    `message #${e.data.no}: ran a task for input (${e.data.input}ms), actual duration: ${taskSimulationDuration.toFixed(2)}ms`
+  );
+  console.log(
+    `message #${e.data.no}: total time from postMessage to task end: ` +
+    `task duration (${taskSimulationDuration.toFixed(2)}) + ` +
+    `message queue wait time etc. (${messageQueueWaitTime.toFixed(2)}) = ` +
+    `${(taskSimulationDuration + messageQueueWaitTime).toFixed(2)} ms (approx)`
+  );
+};
+```
+
+The following timeline illustrates message handling:
+![](timeline_long_task.png)
+
+In this timeline, messages \#1, \#2, and \#3 are handled promptly because their simulated tasks (50ms) complete within the 60ms interval at which messages are sent.
+
+However, message \#4's task is instructed to run for 120ms. While it's processing, message \#5 (sent 60ms after message \#4 was sent) arrives at the worker. Message \#5 must wait in the worker's message queue until message \#4 completes. This results in message \#5 experiencing a significant delay (approximately 60ms) before its handler can even begin.
+
+Manually instrumenting code with `performance.now()` and `event.timeStamp` can help identify the root cause of delays as shown. However, in complex real-world applications, precisely identifying which long task caused a specific message delay, or distinguishing between delay caused by a preceding long task versus a message's own long handler, is very challenging without comprehensive, dedicated monitoring. This API aims to simplify such diagnoses.
+
+## Case 2: Message Queue Congestion
+
+Even without individual long-running tasks, a high volume of messages sent in a short period can overwhelm the receiving context. The message queue can grow, leading to congestion and delays for all subsequent messages, including potentially urgent ones.
+
+The following example codes demonstrates this situation, where multiple `deleteMail` tasks congest the message queue before an urgent `readMail` task.
+
+[Link to live example](https://joone.github.io/web/explainers/delayed_messages/congested/)
+
+### index.html
+
+```js
+<!doctype html>
+<html lang="en">
+  <head>
+    <meta charset="UTF-8" />
+    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
+    <title>An example of a message queue experiencing congestion</title>
+  </head>
+  <body>
+    <h1>IndexedDB Simulation</h1>
+    <button onclick="sendTasksToWorker()">Start</button>
+    <script src="main.js"></script>
+  </body>
+</html>
+```
+### main.js
+
+In main.js, the email application sends 10 deleteEmail tasks every 30 ms to clear junk emails, keeping the worker occupied with intensive processing. Shortly after, the user requests to check their emails, requiring an immediate response.
+
+```js
+// Create a Web Worker
+const worker = new Worker("worker.js");
+
+// Open IndexedDB

```

Comment:
we would generally not recommend using IndexedDB this way --- the API has first class support for simultaneous connections from multiple threads or even processes --- so this is perhaps not the strongest example to lead with.

Reply:
This is not a specific IndexedDB situation, but rather a general scenario where the message queue is congested with numerous tasks. I will update this example to exclude any mention of IndexedDB in order to prevent any confusion.